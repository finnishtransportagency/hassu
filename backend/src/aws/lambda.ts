import { Lambda } from "@aws-sdk/client-lambda";
import { log } from "../logger";

const lambda: Lambda = new Lambda({});

export async function invokeLambda(functionName: string, synchronousCall: boolean, payload?: string): Promise<string | undefined> {
  try {
    if (synchronousCall) {
      const output = await lambda.invoke({
        FunctionName: functionName,
        Payload: new TextEncoder().encode(payload || "{}"),
      });
      const outputPayload = output.Payload;
      return new TextDecoder("utf-8").decode(outputPayload);
    } else {
      await lambda.invokeAsync({
        FunctionName: functionName,
        InvokeArgs: payload || "{}",
      });
    }
  } catch (e) {
    log.error("invokeLambda", e);
    throw e;
  }
}
