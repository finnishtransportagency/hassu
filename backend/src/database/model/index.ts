export * from "./common";
export * from "./projekti";
export * from "./suunnitteluVaihe";
export * from "./nahtavillaoloVaihe";
export * from "./hyvaksymisPaatosVaihe";
