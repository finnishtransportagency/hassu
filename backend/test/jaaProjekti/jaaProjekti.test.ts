import { describe, it } from "mocha";
import * as sinon from "sinon";
import { projektiDatabase } from "../../src/database/projektiDatabase";
import { UserFixture } from "../fixture/userFixture";
import { expect } from "chai";
import { DBProjekti, Palaute, Velho } from "../../src/database/model";
import { jaaProjekti } from "../../src/jaaProjekti";
import { IllegalArgumentError } from "hassu-common/error";
import { userService } from "../../src/user";
import { DBProjektiForSpecificVaiheFixture, VaiheenTila } from "../fixture/DBProjekti2ForSecificVaiheFixture";
import { Kayttaja, Vaihe } from "hassu-common/graphql/apiModel";
import { cloneDeep } from "lodash";
import { velho as velhoClient } from "../../src/velho/velhoClient";
import {
  DynamoDBDocumentClient,
  DynamoDBDocumentClientResolvedConfig,
  PutCommandOutput,
  ServiceInputTypes as DynamoDBServiceInputTypes,
  ServiceOutputTypes as DynamoDBServiceOutputTypes,
  TransactWriteCommand,
  UpdateCommand,
} from "@aws-sdk/lib-dynamodb";

import { AwsStub, mockClient } from "aws-sdk-client-mock";
import {
  CopyObjectCommand,
  DeleteObjectCommand,
  ListObjectsV2Command,
  S3Client,
  S3ClientResolvedConfig,
  ServiceInputTypes as S3ServiceInputTypes,
  ServiceOutputTypes as S3ServiceOutputTypes,
} from "@aws-sdk/client-s3";
import { ProjektiPaths } from "../../src/files/ProjektiPath";
import { lisaAineistoService } from "../../src/tiedostot/lisaAineistoService";
import { PersonSearchFixture } from "../personSearch/lambda/personSearchFixture";
import { personSearch } from "../../src/personSearch/personSearchClient";
import { Kayttajas } from "../../src/personSearch/kayttajas";
import { parameters } from "../../src/aws/parameters";
import { DBMuistuttaja, muistuttajaDatabase } from "../../src/database/muistuttajaDatabase";
import { feedbackDatabase } from "../../src/database/palauteDatabase";
import { uuid } from "hassu-common/util/uuid";

describe("jaaProjekti", () => {
  const userFixture = new UserFixture(userService);

  afterEach(() => {
    userFixture.logout();
    sinon.restore();
  });

  it("should throw error if user is not authenticted", async () => {
    await expect(jaaProjekti({ oid: "oid-123", versio: 1, targetOid: "toinen-oid-234" })).to.eventually.be.rejectedWith(
      Error,
      "Väylä-kirjautuminen puuttuu"
    );
  });

  it("should throw error if user not admin", async () => {
    userFixture.loginAs(UserFixture.mattiMeikalainen);
    await expect(jaaProjekti({ oid: "oid-123", versio: 1, targetOid: "toinen-oid-234" })).to.eventually.be.rejectedWith(
      Error,
      "Sinulla ei ole admin-oikeuksia"
    );
  });

  it("should throw error if VLS-projekti does not exist with oid", async () => {
    userFixture.loginAs(UserFixture.hassuAdmin);
    const oid = "oid-123";
    sinon.stub(projektiDatabase, "loadProjektiByOid").returns(Promise.resolve(undefined));
    await expect(jaaProjekti({ oid, versio: 1, targetOid: "toinen-oid-234" })).to.eventually.be.rejectedWith(
      IllegalArgumentError,
      `Jaettavaa projektia ei löydy oid:lla '${oid}'`
    );
  });

  it("should throw error if VLS-projekti already exists with targetOid", async () => {
    userFixture.loginAs(UserFixture.hassuAdmin);
    const srcProjekti = new DBProjektiForSpecificVaiheFixture().getProjektiForVaihe(Vaihe.NAHTAVILLAOLO, VaiheenTila.LUONNOS);
    const targetProjekti: DBProjekti = { ...cloneDeep(srcProjekti), oid: "toinen-oid" };
    sinon
      .stub(projektiDatabase, "loadProjektiByOid")
      .withArgs(srcProjekti.oid)
      .returns(Promise.resolve(srcProjekti))
      .withArgs(targetProjekti.oid)
      .returns(Promise.resolve(targetProjekti));
    await expect(
      jaaProjekti({ oid: srcProjekti.oid, versio: srcProjekti.versio, targetOid: targetProjekti.oid })
    ).to.eventually.be.rejectedWith(IllegalArgumentError, `Kohde projekti oid:lla '${targetProjekti.oid}' on jo VLS-järjestelmässä`);
  });

  it("should throw error if Velho-projekti does not exist with targetOid", async () => {
    userFixture.loginAs(UserFixture.hassuAdmin);
    const srcProjekti = new DBProjektiForSpecificVaiheFixture().getProjektiForVaihe(Vaihe.NAHTAVILLAOLO, VaiheenTila.LUONNOS);
    const targetProjektiOid = "toinen-oid";
    sinon
      .stub(projektiDatabase, "loadProjektiByOid")
      .withArgs(srcProjekti.oid)
      .returns(Promise.resolve(srcProjekti))
      .withArgs(targetProjektiOid)
      .returns(Promise.resolve(undefined));
    const targetProjektiFromVelho: DBProjekti = {
      oid: targetProjektiOid,
      versio: 1,
      kayttoOikeudet: [],
    };
    sinon.stub(velhoClient, "loadProjekti").withArgs(targetProjektiOid).returns(Promise.resolve(targetProjektiFromVelho));
    await expect(
      jaaProjekti({ oid: srcProjekti.oid, versio: srcProjekti.versio, targetOid: targetProjektiOid })
    ).to.eventually.be.rejectedWith(Error, `Kohde projektia oid:lla '${targetProjektiOid}' ei löydy Projektivelhosta`);
  });

  describe("with admin user and correct parameters", () => {
    let srcProjekti: DBProjekti;
    const targetProjektiOid = "toinen-oid";
    let createProjektiStub: sinon.SinonStub<[projekti: DBProjekti], Promise<PutCommandOutput>>;
    let dynamoDBClient: AwsStub<DynamoDBServiceInputTypes, DynamoDBServiceOutputTypes, DynamoDBDocumentClientResolvedConfig>;

    let targetVelho: Velho;
    let targetProjektiFromVelho: DBProjekti;
    let s3Mock: AwsStub<S3ServiceInputTypes, S3ServiceOutputTypes, S3ClientResolvedConfig>;
    let uuidStub: sinon.SinonStub<[], string>;

    let getKayttajasStub: sinon.SinonStub;
    let a1User: Kayttaja;
    let a2User: Kayttaja;
    let x1User: Kayttaja;
    let srcMuistuttajat: DBMuistuttaja[];
    let srcFeedback: Palaute[];

    beforeEach(() => {
      sinon.stub(parameters, "isAsianhallintaIntegrationEnabled").returns(Promise.resolve(false));
      sinon.stub(parameters, "isUspaIntegrationEnabled").returns(Promise.resolve(false));
      uuidStub = sinon.stub(uuid, "v4");
      const personSearchFixture = new PersonSearchFixture();
      a1User = personSearchFixture.createKayttaja("A1");
      a2User = personSearchFixture.createKayttaja("A2");
      x1User = personSearchFixture.createKayttaja("X1");
      getKayttajasStub = sinon.stub(personSearch, "getKayttajas");
      getKayttajasStub.resolves(Kayttajas.fromKayttajaList([a1User, a2User, x1User]));
      userFixture.loginAs(UserFixture.hassuAdmin);
      srcProjekti = new DBProjektiForSpecificVaiheFixture().getProjektiForVaihe(Vaihe.NAHTAVILLAOLO, VaiheenTila.LUONNOS);
      createProjektiStub = sinon.stub(projektiDatabase, "createProjekti");
      sinon
        .stub(projektiDatabase, "loadProjektiByOid")
        .withArgs(srcProjekti.oid)
        .returns(Promise.resolve(srcProjekti))
        .withArgs(targetProjektiOid)
        .returns(Promise.resolve(undefined));
      sinon.stub(lisaAineistoService, "generateSalt").returns("foo-salt");
      targetVelho = { ...cloneDeep(srcProjekti.velho), nimi: "Toinen-oid projektin nimi" };
      targetProjektiFromVelho = {
        oid: targetProjektiOid,
        versio: 1,
        velho: targetVelho,
        kayttoOikeudet: [],
      };
      sinon.stub(velhoClient, "loadProjekti").withArgs(targetProjektiOid).returns(Promise.resolve(targetProjektiFromVelho));
      s3Mock = mockClient(S3Client);
      s3Mock
        .on(ListObjectsV2Command, {
          Prefix: "yllapito/tiedostot/projekti/2",
          Bucket: "hassu-localstack-yllapito",
        })
        .resolves({
          Contents: [
            { Key: new ProjektiPaths(srcProjekti.oid).sijaintitieto().yllapitoFullPath },
            { Key: "yllapito/tiedostot/projekti/2/syva/tiedosto/polku/tiedosto.json" },
          ],
        });
      s3Mock.on(CopyObjectCommand).resolves({});
      dynamoDBClient = mockClient(DynamoDBDocumentClient);
      srcMuistuttajat = [{ id: "123", expires: 123, lisatty: "2022-01-01", oid: srcProjekti.oid }];
      srcFeedback = [{ id: "123", vastaanotettu: "2022-01-01", oid: srcProjekti.oid }];
      sinon.stub(muistuttajaDatabase, "haeProjektinKaytossaolevatMuistuttajat").returns(Promise.resolve(srcMuistuttajat));
      sinon.stub(feedbackDatabase, "listFeedback").returns(Promise.resolve(srcFeedback));
    });

    afterEach(() => {
      s3Mock.reset();
    });

    it("should copy projekti's fields and mark each julkaisu as copied", async () => {
      await expect(jaaProjekti({ oid: srcProjekti.oid, versio: srcProjekti.versio, targetOid: targetProjektiOid })).to.eventually.be
        .fulfilled;
      expect(createProjektiStub.calledOnce).to.be.true;
      const targetProjektiToCreate = createProjektiStub.firstCall.args[0];
      expect(targetProjektiToCreate).toMatchSnapshot();
      expect(targetProjektiToCreate.oid).to.equal(targetProjektiOid);
      expect(targetProjektiToCreate.velho).to.eql(targetVelho);
      expect(targetProjektiToCreate.salt).to.equal("foo-salt");
      expect(targetProjektiToCreate.kayttoOikeudet).to.eql([
        {
          elyOrganisaatio: undefined,
          email: "pekka.projari@vayla.fi",
          etunimi: "Pekka",
          kayttajatunnus: "A123",
          muokattavissa: false,
          organisaatio: "Väylävirasto",
          puhelinnumero: "123456789",
          sukunimi: "Projari",
          tyyppi: "PROJEKTIPAALLIKKO",
        },
        {
          elyOrganisaatio: undefined,
          email: "Matti.Meikalainen@vayla.fi",
          etunimi: "Matti",
          kayttajatunnus: "A000111",
          muokattavissa: true,
          organisaatio: "Väylävirasto",
          puhelinnumero: "123456789",
          sukunimi: "Meikalainen",
        },
      ]);
      expect(targetProjektiToCreate.aloitusKuulutusJulkaisut?.length).to.equal(1);
      expect(targetProjektiToCreate.aloitusKuulutusJulkaisut?.[0].kopioituProjektista).to.equal(srcProjekti.oid);
      expect(targetProjektiToCreate.vuorovaikutusKierrosJulkaisut?.length).to.equal(1);
      expect(targetProjektiToCreate.vuorovaikutusKierrosJulkaisut?.[0].kopioituProjektista).to.equal(srcProjekti.oid);
      expect(targetProjektiToCreate.nahtavillaoloVaiheJulkaisut).to.be.undefined;
      expect(targetProjektiToCreate.projektinJakautuminen?.jaettuProjektista).to.eql(srcProjekti.oid);
      expect(targetProjektiToCreate.projektinJakautuminen?.jaettuProjekteihin).to.eql(undefined);

      const updateCommands = dynamoDBClient.commandCalls(UpdateCommand);
      expect(updateCommands.length).to.equal(1);
      const input = updateCommands[0].args[0].input;
      expect(input).to.eql({
        ConditionExpression: "(attribute_not_exists(versio) OR versio = :versio) AND attribute_not_exists(projektinJakautuminen)",
        ExpressionAttributeValues: {
          ":projektinJakautuminen": {
            jaettuProjekteihin: [targetProjektiOid],
          },
          ":one": 1,
          ":versio": 1,
        },
        Key: {
          oid: "2",
        },
        TableName: "Projekti-localstack",
        UpdateExpression: "ADD versio :one SET projektinJakautuminen = :projektinJakautuminen",
      });
    });

    it("should copy all srcProjekti's s3objects to targetProjekti's path", async () => {
      await expect(jaaProjekti({ oid: srcProjekti.oid, versio: srcProjekti.versio, targetOid: targetProjektiOid })).to.eventually.be
        .fulfilled;
      expect(s3Mock.calls().length).to.equal(4);
      const listObjectCommands = s3Mock.commandCalls(ListObjectsV2Command);
      expect(listObjectCommands.length).to.equal(1);
      expect(listObjectCommands[0].args[0].input).to.eql({
        Prefix: `yllapito/tiedostot/projekti/${srcProjekti.oid}`,
        Bucket: "hassu-localstack-yllapito",
      });
      const copyObjectCommands = s3Mock.commandCalls(CopyObjectCommand);
      expect(copyObjectCommands.length).to.equal(2);
      expect(copyObjectCommands[0].args[0].input).to.eql({
        Bucket: "hassu-localstack-yllapito",
        ChecksumAlgorithm: "CRC32",
        CopySource: "hassu-localstack-yllapito%2Fyllapito%2Ftiedostot%2Fprojekti%2F2%2Fsijaintitieto",
        Key: "yllapito/tiedostot/projekti/toinen-oid/sijaintitieto",
      });
      expect(copyObjectCommands[1].args[0].input).to.eql({
        Bucket: "hassu-localstack-yllapito",
        ChecksumAlgorithm: "CRC32",
        CopySource: "hassu-localstack-yllapito%2Fyllapito%2Ftiedostot%2Fprojekti%2F2%2Fsyva%2Ftiedosto%2Fpolku%2Ftiedosto.json",
        Key: "yllapito/tiedostot/projekti/toinen-oid/syva/tiedosto/polku/tiedosto.json",
      });
      const deleteObjectCommands = s3Mock.commandCalls(DeleteObjectCommand);
      expect(deleteObjectCommands[0].args[0].input).to.eql({
        Bucket: "hassu-localstack-yllapito",
        Key: "yllapito/tiedostot/projekti/toinen-oid/karttarajaus/karttarajaus.geojson",
      });
    });

    it("should copy muistuttajat and palautteet", async () => {
      uuidStub.onFirstCall().returns("1");
      uuidStub.onSecondCall().returns("2");
      await expect(jaaProjekti({ oid: srcProjekti.oid, versio: srcProjekti.versio, targetOid: targetProjektiOid })).to.eventually.be
        .fulfilled;
      const updateCommands = dynamoDBClient.commandCalls(TransactWriteCommand);
      expect(updateCommands.length).to.equal(2);
      const tallennettuMuistuttaja = updateCommands[0].args[0].input.TransactItems?.[0].Put?.Item;
      expect(tallennettuMuistuttaja).to.eql({ ...srcMuistuttajat[0], oid: targetProjektiOid, id: "1" });
      const tallennettuPalaute = updateCommands[1].args[0].input.TransactItems?.[0].Put?.Item;
      expect(tallennettuPalaute).to.eql({ ...srcFeedback[0], oid: targetProjektiOid, id: "2" });
    });
  });
});
