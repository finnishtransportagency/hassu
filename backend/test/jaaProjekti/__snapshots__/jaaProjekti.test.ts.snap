// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jaaProjekti with admin user and correct parameters should copy projekti's fields and mark each julkaisu as copied 1`] = `
Object {
  "aloitusKuulutus": Object {
    "hankkeenKuvaus": Object {
      "RUOTSI": "Designplatsen är en del av elektrifieringsprojektet Hyvinge-Hangö, som också kommer att genomföra plankorsningsåtgärder. I den fastställda spårplanen är borttagandet av tävlingsstoppet och Leksvallskorsningen genom förbättring av en ny plankorsning vid Helmströms plankorsning vid denna punkt godkänd som planlösning. En uppdatering av Spårplanen kommer nu att lanseras då och man planerar att ta bort alla tre plankorsningarna på den nya Leksvallsöverfarten.",
      "SUOMI": "Suunnittelu kohde on osa Hyvinkää-Hanko sähköistyshanketta, jossa toteutetaan myös tasoristeysten toimenpiteitä. Hyväksytyssä ratasuunnitelmassa kyseisessä kohdassa on hyväksytty suunnitelmaratkaisuna Kisan seisakkeen ja Leksvallin tasoristeysten poistaminen parantamalla Helmströmin tasoristeyksen kohdalle uusi tasoristeys. Nyt käynnistetään Rata-suunnitelman päivitys kyseisessä kohdassa ja suunnitellaan kaikkien kolmen tasoristeyksen poistaminen uudella Leksvallin ylikulkusillalla.",
    },
    "id": 1,
    "ilmoituksenVastaanottajat": Object {
      "kunnat": Array [
        Object {
          "id": 491,
          "sahkoposti": "mikkeli@mikke.li",
        },
        Object {
          "id": 178,
          "sahkoposti": "juva@ju.va",
        },
        Object {
          "id": 740,
          "sahkoposti": "savonlinna@savonlin.na",
        },
      ],
      "viranomaiset": Array [
        Object {
          "nimi": "ETELA_SAVO_ELY",
          "sahkoposti": "kirjaamo.etela-savo@ely-keskus.fi",
        },
      ],
    },
    "kuulutusPaiva": "2022-03-28T14:28",
    "kuulutusYhteystiedot": Object {
      "yhteysHenkilot": Array [
        "A123",
        "A000111",
      ],
      "yhteysTiedot": Array [
        Object {
          "etunimi": "Marko",
          "organisaatio": "Kajaani",
          "puhelinnumero": "0293121213",
          "sahkoposti": "markku.koi@koi.com",
          "sukunimi": "Koi",
        },
      ],
    },
    "siirtyySuunnitteluVaiheeseen": "2022-04-28T14:28",
  },
  "aloitusKuulutusJulkaisut": Array [
    Object {
      "aloituskuulutusPDFt": Object {
        "RUOTSI": Object {
          "aloituskuulutusIlmoitusPDFPath": "/aloituskuulutus/1/MEDDELANDE OM KUNGORELSE FRAN BEHORIG MYNDIGHET Marikas testprojekt.pdf",
          "aloituskuulutusPDFPath": "/aloituskuulutus/1/KUNGORELSE OM INLEDANDET AV PLANERINGEN Marikas testprojekt.pdf",
        },
        "SUOMI": Object {
          "aloituskuulutusIlmoitusPDFPath": "/aloituskuulutus/1/ILMOITUS TOIMIVALTAISEN VIRANOMAISEN KUULUTUKSESTA Marikan testiprojekti.pdf",
          "aloituskuulutusPDFPath": "/aloituskuulutus/1/KUULUTUS SUUNNITTELUN ALOITTAMISESTA Marikan testiprojekti.pdf",
        },
      },
      "hankkeenKuvaus": Object {
        "RUOTSI": "Designplatsen är en del av elektrifieringsprojektet Hyvinge-Hangö, som också kommer att genomföra plankorsningsåtgärder. I den fastställda spårplanen är borttagandet av tävlingsstoppet och Leksvallskorsningen genom förbättring av en ny plankorsning vid Helmströms plankorsning vid denna punkt godkänd som planlösning. En uppdatering av Spårplanen kommer nu att lanseras då och man planerar att ta bort alla tre plankorsningarna på den nya Leksvallsöverfarten.",
        "SUOMI": "Suunnittelu kohde on osa Hyvinkää-Hanko sähköistyshanketta, jossa toteutetaan myös tasoristeysten toimenpiteitä. Hyväksytyssä ratasuunnitelmassa kyseisessä kohdassa on hyväksytty suunnitelmaratkaisuna Kisan seisakkeen ja Leksvallin tasoristeysten poistaminen parantamalla Helmströmin tasoristeyksen kohdalle uusi tasoristeys. Nyt käynnistetään Rata-suunnitelman päivitys kyseisessä kohdassa ja suunnitellaan kaikkien kolmen tasoristeyksen poistaminen uudella Leksvallin ylikulkusillalla.",
      },
      "hyvaksyja": "A123",
      "hyvaksymisPaiva": "2022-03-21",
      "id": 1,
      "ilmoituksenVastaanottajat": Object {
        "kunnat": Array [
          Object {
            "id": 491,
            "lahetetty": "2022-03-11T14:54",
            "sahkoposti": "mikkeli@mikke.li",
          },
          Object {
            "id": 178,
            "lahetetty": "2022-03-11T14:54",
            "sahkoposti": "juva@ju.va",
          },
          Object {
            "id": 740,
            "lahetetty": "2022-03-11T14:54",
            "sahkoposti": "savonlinna@savonlin.na",
          },
        ],
        "viranomaiset": Array [
          Object {
            "lahetetty": "2022-03-11T14:54",
            "nimi": "ETELA_SAVO_ELY",
            "sahkoposti": "kirjaamo.etela-savo@ely-keskus.fi",
          },
        ],
      },
      "kielitiedot": Object {
        "ensisijainenKieli": "SUOMI",
        "projektinNimiVieraskielella": "Marikas testprojekt",
        "toissijainenKieli": "RUOTSI",
      },
      "kopioituToiseltaProjektilta": true,
      "kuulutusPaiva": "2022-03-28T14:28",
      "kuulutusYhteystiedot": Object {
        "yhteysHenkilot": Array [
          "A123",
          "A000111",
        ],
        "yhteysTiedot": Array [
          Object {
            "etunimi": "Marko",
            "organisaatio": "Kajaani",
            "puhelinnumero": "0293121213",
            "sahkoposti": "markku.koi@koi.com",
            "sukunimi": "Koi",
          },
        ],
      },
      "muokkaaja": "A000111",
      "siirtyySuunnitteluVaiheeseen": "2022-04-28T14:28",
      "tila": "HYVAKSYTTY",
      "velho": Object {
        "asiatunnusELY": "ELY/2/2022",
        "asiatunnusVayla": "VAYLA/2/2022",
        "kunnat": Array [
          491,
          178,
          740,
        ],
        "maakunnat": Array [
          1,
          6,
        ],
        "nimi": "Testiprojekti 2",
        "suunnittelustaVastaavaViranomainen": "UUDENMAAN_ELY",
        "tyyppi": "TIE",
        "vastuuhenkilonEmail": "pekka.projari@vayla.fi",
        "vaylamuoto": Array [
          "tie",
        ],
      },
      "yhteystiedot": Array [
        Object {
          "etunimi": "Marko",
          "organisaatio": "Kajaani",
          "puhelinnumero": "0293121213",
          "sahkoposti": "markku.koi@koi.com",
          "sukunimi": "Koi",
        },
      ],
    },
  ],
  "euRahoitus": false,
  "jaettuProjektista": "2",
  "kasittelynTila": Object {
    "hyvaksymispaatos": Object {
      "asianumero": "traficom-123",
      "paatoksenPvm": "2022-02-03",
    },
  },
  "kayttoOikeudet": Array [],
  "kielitiedot": Object {
    "ensisijainenKieli": "SUOMI",
    "projektinNimiVieraskielella": "Marikas testprojekt",
    "toissijainenKieli": "RUOTSI",
  },
  "nahtavillaoloVaihe": Object {
    "hankkeenKuvaus": Object {
      "SUOMI": "Lorem Ipsum nahtavillaoloVaihe",
    },
    "id": 1,
    "ilmoituksenVastaanottajat": Object {
      "kunnat": Array [
        Object {
          "id": 491,
          "lahetetty": "2022-03-11T14:54",
          "sahkoposti": "mikkeli@mikke.li",
        },
        Object {
          "id": 178,
          "lahetetty": "2022-03-11T14:54",
          "sahkoposti": "juva@ju.va",
        },
        Object {
          "id": 740,
          "lahetetty": "2022-03-11T14:54",
          "sahkoposti": "savonlinna@savonlin.na",
        },
      ],
      "viranomaiset": Array [
        Object {
          "lahetetty": "2022-03-11T14:54",
          "nimi": "ETELA_SAVO_ELY",
          "sahkoposti": "kirjaamo.etela-savo@ely-keskus.fi",
        },
      ],
    },
    "kuulutusPaiva": "2022-06-07",
    "kuulutusVaihePaattyyPaiva": "2022-06-07",
    "kuulutusYhteystiedot": Object {
      "yhteysHenkilot": Array [
        "A123",
        "A000111",
      ],
      "yhteysTiedot": Array [
        Object {
          "etunimi": "Marko",
          "organisaatio": "Kajaani",
          "puhelinnumero": "0293121213",
          "sahkoposti": "markku.koi@koi.com",
          "sukunimi": "Koi",
        },
      ],
    },
    "muistutusoikeusPaattyyPaiva": "2022-06-08",
  },
  "nahtavillaoloVaiheJulkaisut": undefined,
  "oid": "toinen-oid",
  "paivitetty": "2022-03-15T14:30:00.000Z",
  "salt": "foo-salt",
  "tallennettu": true,
  "vahainenMenettely": false,
  "velho": Object {
    "asiatunnusELY": "ELY/2/2022",
    "asiatunnusVayla": "VAYLA/2/2022",
    "kunnat": Array [
      491,
      178,
      740,
    ],
    "maakunnat": Array [
      1,
      6,
    ],
    "nimi": "Toinen-oid projektin nimi",
    "suunnittelustaVastaavaViranomainen": "UUDENMAAN_ELY",
    "tyyppi": "TIE",
    "vastuuhenkilonEmail": "pekka.projari@vayla.fi",
    "vaylamuoto": Array [
      "tie",
    ],
  },
  "versio": 1,
  "vuorovaikutusKierros": Object {
    "arvioSeuraavanVaiheenAlkamisesta": Object {
      "RUOTSI": "Höst 2024",
      "SUOMI": "Syksy 2024",
    },
    "esitettavatYhteystiedot": Object {
      "yhteysHenkilot": Array [
        "A000112",
      ],
      "yhteysTiedot": Array [],
    },
    "hankkeenKuvaus": Object {
      "RUOTSI": "svenska",
      "SUOMI": "Tavoitteena on nykyisen ja tulevan maankäytön liittäminen
luontevasti Hämeenlinnanväylään, huomioida alueen melunsuojaus, parantaa henkilöautoliikenteen ja joukkoliikenteen
sujuvuutta ja turvallisuutta sekä tehdä jalankulun ja pyöräilyn
yhteydet sujuviksi ja turvallisiksi. Raskaan liikenteen sujuvuuden ja matka-ajan ennustettavuuden parantaminen on myös
yksi tavoitteista.",
    },
    "ilmoituksenVastaanottajat": Object {
      "kunnat": Array [
        Object {
          "id": 491,
          "lahetetty": "2022-03-11T14:54",
          "sahkoposti": "mikkeli@mikke.li",
        },
        Object {
          "id": 178,
          "lahetetty": "2022-03-11T14:54",
          "sahkoposti": "juva@ju.va",
        },
        Object {
          "id": 740,
          "lahetetty": "2022-03-11T14:54",
          "sahkoposti": "savonlinna@savonlin.na",
        },
      ],
      "viranomaiset": Array [
        Object {
          "lahetetty": "2022-03-11T14:54",
          "nimi": "ETELA_SAVO_ELY",
          "sahkoposti": "kirjaamo.etela-savo@ely-keskus.fi",
        },
      ],
    },
    "suunnittelunEteneminenJaKesto": Object {
      "RUOTSI": "RUOTSIKSI Välin Kehä I–Kaivoksela tiesuunnitelma valmistuu 4/2021.
Hankkeen jatkosuunnittelun ja toteuttamisen aikataulusta
ei ole päätöksiä.",
      "SUOMI": "Välin Kehä I–Kaivoksela tiesuunnitelma valmistuu 4/2021.
Hankkeen jatkosuunnittelun ja toteuttamisen aikataulusta
ei ole päätöksiä.",
    },
    "tila": "JULKINEN",
    "vuorovaikutusJulkaisuPaiva": "2022-04-28T14:28",
    "vuorovaikutusNumero": 1,
    "vuorovaikutusTilaisuudet": Array [
      Object {
        "alkamisAika": "15:00",
        "kaytettavaPalvelu": "TEAMS",
        "linkki": "https://linkki_tilaisuuteen",
        "nimi": Object {
          "RUOTSI": "RUOTSIKSI Lorem ipsum",
          "SUOMI": "Lorem ipsum",
        },
        "paattymisAika": "16:00",
        "paivamaara": "2022-03-04",
        "tyyppi": "VERKOSSA",
      },
    ],
  },
  "vuorovaikutusKierrosJulkaisut": Array [
    Object {
      "arvioSeuraavanVaiheenAlkamisesta": Object {
        "RUOTSI": "Höst 2024",
        "SUOMI": "Syksy 2024",
      },
      "esitettavatYhteystiedot": Object {
        "yhteysHenkilot": Array [
          "A000112",
        ],
        "yhteysTiedot": Array [],
      },
      "hankkeenKuvaus": Object {
        "RUOTSI": "svenska",
        "SUOMI": "Tavoitteena on nykyisen ja tulevan maankäytön liittäminen
luontevasti Hämeenlinnanväylään, huomioida alueen melunsuojaus, parantaa henkilöautoliikenteen ja joukkoliikenteen
sujuvuutta ja turvallisuutta sekä tehdä jalankulun ja pyöräilyn
yhteydet sujuviksi ja turvallisiksi. Raskaan liikenteen sujuvuuden ja matka-ajan ennustettavuuden parantaminen on myös
yksi tavoitteista.",
      },
      "id": 1,
      "ilmoituksenVastaanottajat": Object {
        "kunnat": Array [
          Object {
            "id": 491,
            "lahetetty": "2022-03-11T14:54",
            "sahkoposti": "mikkeli@mikke.li",
          },
          Object {
            "id": 178,
            "lahetetty": "2022-03-11T14:54",
            "sahkoposti": "juva@ju.va",
          },
          Object {
            "id": 740,
            "lahetetty": "2022-03-11T14:54",
            "sahkoposti": "savonlinna@savonlin.na",
          },
        ],
        "viranomaiset": Array [
          Object {
            "lahetetty": "2022-03-11T14:54",
            "nimi": "ETELA_SAVO_ELY",
            "sahkoposti": "kirjaamo.etela-savo@ely-keskus.fi",
          },
        ],
      },
      "kopioituToiseltaProjektilta": true,
      "suunnittelunEteneminenJaKesto": Object {
        "RUOTSI": "RUOTSIKSI Välin Kehä I–Kaivoksela tiesuunnitelma valmistuu 4/2021.
Hankkeen jatkosuunnittelun ja toteuttamisen aikataulusta
ei ole päätöksiä.",
        "SUOMI": "Välin Kehä I–Kaivoksela tiesuunnitelma valmistuu 4/2021.
Hankkeen jatkosuunnittelun ja toteuttamisen aikataulusta
ei ole päätöksiä.",
      },
      "tila": "JULKINEN",
      "vuorovaikutusJulkaisuPaiva": "2022-04-28T14:28",
      "vuorovaikutusPDFt": Object {
        "RUOTSI": Object {
          "kutsuPDFPath": "suunnitteluvaihe/vuorovaikutus_1/2.pdf",
        },
        "SUOMI": Object {
          "kutsuPDFPath": "suunnitteluvaihe/vuorovaikutus_1/1.pdf",
        },
      },
      "vuorovaikutusTilaisuudet": Array [
        Object {
          "alkamisAika": "15:00",
          "kaytettavaPalvelu": "TEAMS",
          "linkki": "https://linkki_tilaisuuteen",
          "nimi": Object {
            "RUOTSI": "RUOTSIKSI Lorem ipsum",
            "SUOMI": "Lorem ipsum",
          },
          "paattymisAika": "16:00",
          "paivamaara": "2022-03-04",
          "tyyppi": "VERKOSSA",
        },
      ],
      "yhteystiedot": Array [
        Object {
          "etunimi": "Marko",
          "organisaatio": "Kajaani",
          "puhelinnumero": "0293121213",
          "sahkoposti": "markku.koi@koi.com",
          "sukunimi": "Koi",
        },
      ],
    },
  ],
}
`;
